index.js

```diff
@@ -102,6 +104,57 @@ function fastifyPostgres (fastify, options, next) {
     }
   }
 
+  fastify.addHook('onRoute', routeOptions => {
+    const useTransaction = routeOptions.useTransaction
+
+    const transactionFailedSymbol = Symbol('transactionFailed')
+
+    if (useTransaction) {
+      // This will rollback the transaction if the handler fails at some point
+      const onError = async (req, reply, error) => {
+        req[transactionFailedSymbol] = true
+
+        try {
+          await req.pg.query('ROLLBACK')
+        } catch (err) {
+          // await req.pg.query('ROLLBACK')
+        }
+      }
+
+      const preHandler = async (req, reply) => {
+        const client = await pool.connect()
+        req.pg = client
+        await req.pg.query('BEGIN')
+      }
+
+      // This will commit the transaction (or rollback if that fails) and also always
+      // release the client, regardless of error state or useTransaction value
+      const onSend = async (req, reply, payload) => {
+        try {
+          if (!req[transactionFailedSymbol]) {
+            await req.pg.query('COMMIT')
+          }
+        } catch (err) {
+          await req.pg.query('ROLLBACK')
+        } finally {
+          req.pg.release()
+        }
+      }
+
+      // Add these handlers
+      routeOptions.preHandler = addHandler(routeOptions.preHandler, preHandler)
+      routeOptions.onSend = addHandler(routeOptions.onSend, onSend)
+      routeOptions.onError = addHandler(routeOptions.onError, onError)
+    }
+  })
+
+  if (!fastify.hasRequestDecorator('pg')) {
+    fastify.decorateRequest('pg', null)
+    fastify.addHook('onRequest', async (req, reply) => {
+      req.pg = fastify.pg
+    })
+  }
```

Comment:
let remove this for the time being

Reply:
if I were to write this code, I would keep it like: (not 100% sure though)

```suggestion
  if (!fastify.hasRequestDecorator('pg')) {
    fastify.decorateRequest('pg', fastify.pg)
  }
```